frame/gilt/src/lib.rs

```diff
@@ -0,0 +1,578 @@
+		/// \[ who, amount, duration \]
+		BidRetracted(T::AccountId, BalanceOf<T>, u32),
+		/// A bid was accepted as a gilt. The balance may not be released until expiry.
+		/// \[ index, expiry, who, amount \]
+		GiltIssued(ActiveIndex, T::BlockNumber, T::AccountId, BalanceOf<T>),
+		/// An expired gilt has been thawed.
+		/// \[ index, who, original_amount, additional_amount \]
+		GiltThawed(ActiveIndex, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The duration of the bid is less than one.
+		DurationTooSmall,
+		/// The duration is the bid is greater than the number of queues.
+		DurationTooBig,
+		/// The amount of the bid is less than the minimum allowed.
+		AmountTooSmall,
+		/// The queue for the bid's duration is already full.
+		QueueFull,
+		/// Gilt index is unknown.
+		Unknown,
+		/// Not the owner of the gilt.
+		NotOwner,
+		/// Gilt not yet at expiry date.
+		NotExpired,
+		/// The given bid for retraction is not found.
+		NotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if (n % T::IntakePeriod::get()).is_zero() {
+				Self::pursue_target(T::MaxIntakeBids::get())
+			} else {
+				0
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T:Config> Pallet<T> {
+		/// Place a bid for a gilt to be issued.
+		///
+		/// Origin must be Signed, and account must have at least `amount` in free balance.
+		///
+		/// - `amount`: The amount of the bid; these funds will be reserved. If the bid is
+		/// successfully elevated into an issued gilt, then these funds will continue to be
+		/// reserved until the gilt expires. Must be at least `MinFreeze`.
+		/// - `duration`: The number of periods for which the funds will be locked if the gilt is
+		/// issued. It will expire only after this period has elapsed after the point of issuance.
+		/// Must be greater than 1 and no more than `QueueCount`.
+		///
+		/// Complexities:
+		/// - `Queues[duration].len()` (just take max).
+		#[pallet::weight(T::WeightInfo::place_bid_max())]
+		pub fn place_bid(
+			origin: OriginFor<T>,
+			#[pallet::compact] amount: BalanceOf<T>,
+			duration: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(amount >= T::MinFreeze::get(), Error::<T>::AmountTooSmall);
+			let queue_count = T::QueueCount::get() as usize;
+			let queue_index = duration.checked_sub(1)
+				.ok_or(Error::<T>::DurationTooSmall)? as usize;
+			ensure!(queue_index < queue_count, Error::<T>::DurationTooBig);
+
+			let bid = GiltBid { amount, who: who.clone() };
+			let net = Queues::<T>::try_mutate(duration, |q|
+				-> Result<(u32, BalanceOf::<T>), DispatchError>
+			{
+				// queue is <Ordered: Lowest ... Highest><Fifo: Last ... First>
+				let net = if q.len() == T::MaxQueueLen::get() as usize {
+					ensure!(q[0].amount < amount, Error::<T>::QueueFull);
```

Comment:
This looks more it should be `BidAmountTooLow` error or something. `QueueFull` to me implies nothing can be done, but if they simply increase their amount, they could kick out the next lowest.